### DISCUSSION TITLE
Generic short declaration

### DISCUSSION THREAD

**Discussion by denisenko93** [dotnet/csharplang#5173] [Language Ideas]
If a generic type is used in a where, it does not need to be declared in the list of generic types.

Current implementation:

Example. From library MediatR:
```
public interface IRequest<out TResponse>
{
}
public interface IRequestHandler<in TRequest, TResponse> where TRequest : IRequest<TResponse>
{
   Task<TResponse> Handle(TRequest request, CancellationToken cancellationToken);
}
```
Usage:
```
public class SampleResponseBase
{
}

public class SampleResponse: SampleResponseBase
{
}

public class SampleRequest : IRequest<SampleResponse>
{
}

public class SampleHandler : IRequestHandler<SampleRequest, SampleResponseBase>
{ 
	public Task<SampleResponseBase> Handle(SampleRequest request, CancellationToken cancellationToken)
	{
		throw new NotImplementedException();
	}
}
```

In implementation SampleHandler I can use SampleResponse or base types (SampleResponseBase in example) in generic type

New implementation:
```
public interface IRequestHandler<in TRequest> where TRequest : IRequest<TResponse>
{
   Task<TResponse> Handle(TRequest request, CancellationToken cancellationToken);
}

public class SampleHandler : IRequestHandler<SampleRequest>
{ 
	public Task<SampleResponse> Handle(SampleRequest request, CancellationToken cancellationToken)
	{
		throw new NotImplementedException();
	}
}
```

If in generic types not declared type used in where statement, use type from generic declaration in implementation

**Comment by HaloFour** [comment]
See Also: https://github.com/dotnet/csharplang/issues/1328

There have been other discussions about associated types as well.  It may fall out as a part of shapes.

**Comment by acaly** [comment]
I think one reason for this to be difficult is that a type can declare multiple interfaces so when determining the implicit generic argument there will be ambiguity.

I agree that the generic argument inference can be improved, and there are also proposals related to that #1349.
